acceptance_testing.go

```diff
@@ -0,0 +1,834 @@
+				delete(haveCfg, name)
+
+				is.Equal(len(haveCfg)+1, len(origCfg)) // destination config does not contain required parameter, please check the test setup
+
+				dest := a.driver.Connector().NewDestination()
+				err := dest.Configure(ctx, haveCfg)
+				is.True(err != nil)
+
+				err = dest.Teardown(ctx)
+				is.NoErr(err)
+			})
+		}
+	}
+}
+
+func (a acceptanceTest) TestDestination_Write_Success(t *testing.T) {
+	a.skipIfNoDestination(t)
+	is := is.New(t)
+	ctx := context.Background()
+	defer goleak.VerifyNone(t, a.driver.GoleakOptions(t)...)
+
+	dest, cleanup := a.openDestination(ctx, t)
+	defer cleanup()
+
+	want := a.generateRecords(20)
+	var gotSynchronous []Record
+	var gotAsynchronous []Record
+
+	t.Run("synchronous", func(t *testing.T) {
+		is := is.New(t)
+		for i, r := range want {
+			writeCtx, cancel := context.WithTimeout(ctx, time.Second*5)
+			defer cancel()
+			err := dest.Write(writeCtx, r)
+			if i == 0 && errors.Is(err, ErrUnimplemented) {
+				t.Skip("Write not implemented")
+			}
+			is.NoErr(err)
+		}
+
+		// Flush is optional, we allow it to be unimplemented
+		err := dest.Flush(ctx)
+		if !errors.Is(err, ErrUnimplemented) {
+			is.NoErr(err)
+		}
+
+		got := make([]Record, 0, len(want))
+		a.driver.Read(t, &got)
+		gotSynchronous = got
+
+		is.Equal(len(got), len(want)) // destination didn't write expected number of records
+		for i := range want {
+			want[i].Position = got[i].Position   // position can't be determined in advance
+			want[i].CreatedAt = got[i].CreatedAt // created at can't be determined in advance
+
+			is.Equal(want[i], got[i])
+		}
+	})
+	t.Run("asynchronous", func(t *testing.T) {
+		is := is.New(t)
+		var ackWg sync.WaitGroup
+		for i, r := range want {
+			writeCtx, cancel := context.WithTimeout(ctx, time.Second*5)
+			defer cancel()
+
+			ackWg.Add(1)
+			err := dest.WriteAsync(writeCtx, r, func(err error) error {
+				defer ackWg.Done()
+				return err // TODO check error, but not here, we might not be in the right goroutine
+			})
+			if i == 0 && errors.Is(err, ErrUnimplemented) {
+				t.Skip("WriteAsync not implemented")
+			}
+			is.NoErr(err)
+		}
+
+		err := dest.Flush(ctx)
+		is.NoErr(err)
+
+		// wait for acks to get called
+		// TODO timeout if it takes too long
+		ackWg.Done()
+
+		got := make([]Record, 0, len(want))
+		a.driver.Read(t, &got)
+		// skip records retrieved by synchronous function (unlikely to happen, but let's be thorough)
+		got = got[len(gotSynchronous):]
+		gotAsynchronous = got
+
+		is.Equal(len(got), len(want)) // destination didn't write expected number of records
+		for i := range want {
+			want[i].Position = got[i].Position   // position can't be determined in advance
+			want[i].CreatedAt = got[i].CreatedAt // created at can't be determined in advance
+
+			is.Equal(want[i], got[i])
+		}
+	})
+
+	is.True((len(gotSynchronous) > 0) != (len(gotAsynchronous) > 0)) // either Write or WriteAsync should be implemented and working (not both)
```

Comment:
Can we somehow enforce this in the SDK? It would also be good that we mention this in the docs.

Reply:
If by "enforce" you mean that it would produce a compile-time error, I don't see how we could do that. I'll add a comment to the docs about this 👍

Reply:
Or runtime (e.g. when writing the first record, but that is likely not a good experience). 

Not related to this, it may make sense to split test into three: one for sync, one for async, and one checking if exactly one of the methods is implemented.

Reply:
Split into `TestDestination_WriteOrWriteAsync`, `TestDestination_Write_Success` and `TestDestination_WriteAsync_Success` 👍